sherpa/astro/io/xstable.py

```diff
@@ -0,0 +1,752 @@
+>>> linepos = np.arange(0, 2.5, 0.1)
+>>> minval = linepos[0]
+>>> maxval = linepos[-1]
+
+The model is evaluated over the grid defined by ``elo`` and ``ehi``
+for each element of ``linepos``, which is used to set the
+`sherpa.astro.xspec.XSgaussian.LineE` parameter:
+
+>>> models = []
+>>> for lp in linepos:
+...     mdl.linee = lp
+...     ymodel = mdl(elo, ehi)
+...     models.append(ymodel)
+...
+
+This model has a single interpolated parameter which we call
+``"pos"``.  Note that the delta parameter - here set to 0.01 - is only
+used by Sherpa to decide if the parameter is frozen (value is
+negative) or not, but is used by the optimiser in XSPEC. The values
+field is set to those used to generate the spectral models.
+
+>>> param = xstable.IntParam("pos", 1, 0.01, minval, maxval,
+...                          values=linepos)
+
+With this we can create the necessary information to create the
+table model (`make_xstable_model`) and then write it out
+as a FITS file (`write_xstable_model`):
+
+>>> hdus = xstable.make_xstable_model("gaussy", elo, ehi, params=[param],
+...                                   spectra=models)
+>>> xstable.write_xstable_model("example.mod", hdus, clobber=True)
+
+This file can then be used in Sherpa with either
+`sherpa.astro.ui.load_xstable_model` or
+`sherpa.astro.xspec.read_xstable_model`. For example:
+
+>>> from sherpa.astro import ui
+>>> ui.load_xstable_model("mygauss", "example.mod")
+>>> print(mygauss)
+xstablemodel.mygauss
+   Param        Type          Value          Min          Max      Units
+   -----        ----          -----          ---          ---      -----
+   mygauss.pos  thawed            1            0          2.4
+   mygauss.norm thawed            1            0        1e+24
+
+Notes
+-----
+
+XSPEC models can be created without XSPEC support in Sherpa, but it is
+needed to read the files in.
+
+For additive models it is assumed that the model values - that is,
+each bin - have units of photon/cm^2/s. This is easy to accidentally
+change - e.g.  in the example above if mdl.norm were changed to a
+value other than 1 everything would work but Sherpa and XSPEC would
+infer incorrect fluxes or luminosities.
+
+"""
+
+from dataclasses import dataclass, field
+import itertools
+import time
+from typing import Any, Optional, Union
+
+import numpy as np
+
+import sherpa
+
+
+__all__ = ("IntParam", "Param", "make_xstable_model",
+           "write_xstable_model")
+
+
+# In the typing support I have used Any to mean "ndarray".
+#
+
+@dataclass
+class Param:
+    """Represent a parameter.
+
+    The absolute DELTA value is used by XSPEC but ignored by Sherpa.
+    The sign of DELTA is used (by both Sherpa and XSPEC) to decide if
+    the parameter should be frozen (DELTA is negative) or not by
+    default.
+
+    """
+    name: str
+    """The parameter name.
+
+    This must be unique and ideally a valid attribute name in Python,
+    although the latter is not enforced.
+    """
+    initial: float
```

Comment:
I like an empty line betwene the preceeding docsting and the next attribute. I think it makes the code easier to read. Not a requirement though.

Reply:
Actually, as I was editing this file I've added spaces.